boa/src/builtins/json/mod.rs

```diff
@@ -145,133 +148,542 @@ impl Json {
+                            property_set.insert(s.to_owned());
+                        } else if v.is_number() {
+                            property_set.insert(
+                                v.to_string(context)
+                                    .expect("ToString cannot fail on number value"),
+                            );
+                        } else if let Some(obj) = v.as_object() {
+                            // i. If v has a [[StringData]] or [[NumberData]] internal slot, set item to ? ToString(v).
+                            if obj.is_string() || obj.is_number() {
+                                property_set.insert(v.to_string(context)?);
+                            }
+                        }
+
+                        // h. Set k to k + 1.
+                        k += 1;
+                    }
+                    property_list = Some(property_set.into_iter().collect());
+                }
+            }
+        }
+
+        let mut space = args.get(2).cloned().unwrap_or_default();
+
+        // 5. If Type(space) is Object, then
+        if let Some(space_obj) = space.as_object() {
+            // a. If space has a [[NumberData]] internal slot, then
+            if space_obj.is_number() {
+                // i. Set space to ? ToNumber(space).
+                space = space.to_number(context)?.into();
+            }
+            // b. Else if space has a [[StringData]] internal slot, then
+            else if space_obj.is_string() {
+                // i. Set space to ? ToString(space).
+                space = space.to_string(context)?.into();
+            }
+        }
+
+        // 6. If Type(space) is Number, then
+        let gap = if space.is_number() {
+            // a. Let spaceMV be ! ToIntegerOrInfinity(space).
+            // b. Set spaceMV to min(10, spaceMV).
+            // c. If spaceMV < 1, let gap be the empty String; otherwise let gap be the String value containing spaceMV occurrences of the code unit 0x0020 (SPACE).
+            match space
+                .to_integer_or_infinity(context)
+                .expect("ToIntegerOrInfinity cannot fail on number")
+            {
+                IntegerOrInfinity::PositiveInfinity => JsString::new("          "),
+                IntegerOrInfinity::NegativeInfinity => JsString::new(""),
+                IntegerOrInfinity::Integer(i) if i < 1 => JsString::new(""),
+                IntegerOrInfinity::Integer(i) => {
+                    let mut s = String::new();
+                    let i = std::cmp::min(10, i);
+                    for _ in 0..i {
+                        s.push(' ');
+                    }
+                    s.into()
+                }
+            }
+        // 7. Else if Type(space) is String, then
+        } else if let Some(s) = space.as_string() {
+            // a. If the length of space is 10 or less, let gap be space; otherwise let gap be the substring of space from 0 to 10.
+            if s.len() <= 10 {
```

Comment:
Wouldn't this count the number of bytes in `s`? I think this should be encoded to utf_16 before getting its length.

Reply:
True. Nice catch!